src/main/java/com/amazon/opendistroforelasticsearch/security/auditlog/impl/AuditLogImpl.java

```diff
@@ -38,179 +53,468 @@
+        if (enabled) {
+            messageRouter.route(msg);
+        }
+    }
+
+    @Override
+    public void logFailedLogin(final String effectiveUser,
+                               final boolean securityAdmin,
+                               final String initiatingUser,
+                               final TransportRequest request,
+                               final Task task) {
+        if (!enabled) return;
+        if (!AuditFilter.checkTransportFilter(AuditCategory.FAILED_LOGIN, null, effectiveUser, request, auditConfig)) {
+            return;
+        }
+
+        List<AuditMessage> msgs = requestResolver.resolve(AuditCategory.FAILED_LOGIN, getOrigin(), null, null,
+                effectiveUser, securityAdmin, initiatingUser, getRemoteAddress(), request, task, auditConfig, null);
+
+        msgs.forEach(this::save);
+    }
+
+    @Override
+    public void logFailedLogin(final String effectiveUser,
+                               final boolean securityAdmin,
+                               final String initiatingUser,
+                               final RestRequest request) {
+        if (!enabled) return;
+        if (!AuditFilter.checkRestFilter(AuditCategory.FAILED_LOGIN, effectiveUser, request, auditConfig)) {
+            return;
+        }
+
+        AuditMessage msg = new AuditMessageBuilder(AuditCategory.FAILED_LOGIN)
+                .addClusterServiceInfo(clusterService)
+                .addOrigin(getOrigin())
+                .addLayer(Origin.REST)
+                .addRemoteAddress(getRemoteAddress())
+                .addRequestInfo(request, auditConfig.shouldExcludeSensitiveHeaders())
+                .addRequestBody(request, auditConfig.shouldLogRequestBody())
+                .addInitiatingUser(initiatingUser)
+                .addEffectiveUser(effectiveUser)
+                .addIsAdminDn(securityAdmin)
+                .build();
+
+        save(msg);
+    }
+
+    @Override
+    public void logSucceededLogin(final String effectiveUser,
+                                  final boolean securityAdmin,
+                                  final String initiatingUser,
+                                  final TransportRequest request,
+                                  final String action,
+                                  final Task task) {
+        if (!enabled) return;
+        if (!AuditFilter.checkTransportFilter(AuditCategory.AUTHENTICATED, action, effectiveUser, request, auditConfig)) {
+            return;
+        }
+
+        List<AuditMessage> msgs = requestResolver.resolve(AuditCategory.AUTHENTICATED, getOrigin(), action, null,
+                effectiveUser, securityAdmin, initiatingUser, getRemoteAddress(), request, task, auditConfig, null);
+
+        msgs.forEach(this::save);
+    }
+
+    @Override
+    public void logSucceededLogin(String effectiveUser, boolean securityAdmin, String initiatingUser, RestRequest request) {
+        if (!enabled) return;
```

Comment:
any advantage of having this check for each method? Save function is already checking for enabled

Reply:
This would help in early exit instead of processing/creating audit message and then failing.